### ISSUE TITLE
loadmat gives segmentation error or hangs indefinitely

### ISSUE THREAD

**Issue by cachitas** [frejanordsiek/hdf5storage#42]
I'm trying to read a mat file from MATLAB 2016a.

Any ideas on how to get more info on what's happening?

**Comment by frejanordsiek** [comment]
There are a couple possibilities. Thankfully, Mathworks adding a new HDF5 based MAT file format is not one of them https://www.mathworks.com/help/matlab/ref/save.html . R2016a doesn't have any format newer than v7.3.

1. A variable saved from Matlab is not compatible with the hdf5storage package. This would be any class. Sparses, while not supported yet, should at least be read as Numpy structured ndarrays. Matlab classes, when saved, have object references. If there is a circular set of references in how the class is stored, then hdf5storage will read them all repeatedly before running into a memory problem. It currently has no way to detect such a circular dependency (such testing should be added, though).
2. The MAT file has a variable that would be read as a structured Numpy ndarray but would have a dtype larger than signed 32-bits due to the sizes of all fields (each less than signed 32-bits) adding up to more than signed 32-bits (last I checked, size checking is only done on each field, not overflow checking on their sum).

Since it sometimes throws an error, the issue is probably not hdf5storage's poor speed performance.

Can you describe what the types and dimensions of the variables you are trying to read are? Also, are you able to read MAT files with a small number of simple variables (small double arrays, small int8 arrays, simple structures, simple cell arrays, etc.)?

**Comment by cachitas** [comment]
Thank you for your deep thoughts on this matter.
I can share one file with you so you can test it too ([behavior.mat.zip](https://github.com/frejanordsiek/hdf5storage/files/221251/behavior.mat.zip)). The file in question is generated by a third-party application ([JAABA](https://github.com/kristinbranson/JAABA)), so in case it is a problem with the data types or reference structure, there's not much we can do.

I have not tested yet your package with simpler MAT files, since when I found it I jumped right away to my problem, but I assume there's no problem with the HDF5 library because both h5py and pytables seem to work fine.

Meanwhile, I think I overcame my issues using PyTables. Nevertheless, I appreciate the work you have here, so I did not want to let this errors pass unnoticed, so you can improve hdf5storage!

**Comment by frejanordsiek** [comment]
That mat file doesn't seem to be a proper zip file. What program did you use to compress it. I figure I should have a look to see if it isn't a problem with this package.

**Comment by cachitas** [comment]
I just compressed the MAT file in an Ubuntu machine because of GitHub size limitations.
Downloaded it now on a Windows and looks like the MAT file is in there.

**Comment by frejanordsiek** [comment]
Was able to read it finally.

I opened the file in hdfview and it definitely has a few MATLAB classes in it. In addition, it has a lot of object references. This last thing is a known thing that this package is very slow at. I tried reading the individual fields of the main struct called ``'x'`` in the file. Some of them are taking a very long time. One didn't end at all so I terminated it. My guess is that there are enough object references for the package to take a very long time and possibly do many allocations in the heap and therefore running into memory problems.

**Comment by frejanordsiek** [comment]
I did come up with one idea that would solve your problem, possibly. How about saving it in the ``'-v7'`` or ``'-v6'`` formats, which can then be read by scipy. At worst, you could load the MAT files into matlab and then write them in the different format.

**Comment by cachitas** [comment]
I overcame this using PyTables. I only opened the issue to let you know of this problem in your package.

**Comment by frejanordsiek** [comment]
Thank you.

I hope to one day improve the performance issues that this package has. Have to do some major ABI and API revision first to make it more doable. Also hope to implement Matlab classes.

**Comment by frejanordsiek** [comment]
I managed to significantly improve the performance. See Issue #55

Implementing Matlab classes will have to wait

**Comment by soamaven** [comment]
This issue is especially prominent in python 3.5 and 3.6

I was able wait long enough for a large `.mat` to load in my py2.7 conda env, but the py3.5 and py3.6 envs segfault

**Comment by frejanordsiek** [comment]
Segfaulting would indicate a more serious issue is at work here.

Though, improving the performance will still be a good thing, so I will try to apply some of the same tricks done for the master branch to the 0.1.x branch (well, the rest other than one trick that is already done).

**Comment by frejanordsiek** [comment]
Putting this here to measure against. Current performance on my laptop using

```
>>> import numpy as np
>>> import hdf5storage
>>> a = np.object_([np.int8(1)]*10000)
>>> %time hdf5storage.write(a, path='/a', matlab_compatible=False, store_python_metadata=False)
>>> %time b = hdf5storage.read(path='/a')
```

which writes and then reads back a 10k element cell-array like object with int8 scalars in each element is

write speed: 869 elements/s (11.5 s)
read speed: 18 elements/s (544 s)

**Comment by frejanordsiek** [comment]
Adding in all the changes in commits 9cbf1be1, 89c356dc, and a8698123 improve the performance a little bit to

write speed: 885 elements/s (11.3 s)
read speed: 19.6 elements/s (511 s)

which isn't very much. I am going to continue to try to improve the read performance. Not all the tricks I used to solve issue #55 will work for the 0.1.x branch since many of them broke compatibility. One of them I was able to partially do with a clever work around.

**Comment by frejanordsiek** [comment]
While reviewing the code as part of the progress towards a 1.0 release, I returned to ``utilities.read_object_array`` and found the cause of the performance issue. The problem is that the code is iterating over an HDF5 Dataset, which is read one element at a time and is therefore very slow even if the shape and other things are only read once. I have changed it to read in the whole reference array first and then flat iterate over it in commit a3e184446b3fa8b3d2f6216bf283e5b9c3a204e0, which increased the performance considerably. The performance is now:

write speed: 1658 elements/s (6.03 s)
read speed: 3215 elements/s (3.11 s)

which is faster than writing like it should be (there is a lot more logic going on in writing than with reading, so reading should in principle be faster).

I think the performance issue is now closed.
